### ISSUE TITLE
Remove Page and Partial class

### ISSUE THREAD

**Issue by Starcounter-Jack** [Starcounter/Starcounter#3202]
We are discussing fixing the Page/Partial problem.
### Problems
1. Implicit Magic clouds the understanding of Starcounter
2. Having to set a special base class that is easy to forget.
3. Separation of concerns. View models can be used by many types of views and should not provide HTML functionality. Our friend Martin already have a native Windows clients consuming view models.
### Consequences

This will affect the following:
1. HTML merger. Needs to be moved to middleware. It can no longer make use of the `Html` property, so it needs to ask for the HTML rather than the URL.
2. Automatic standalone mode. Needs to be moved to middleware. I suggest that this is not implicit. I.e. the user needs to opt-in using something like `Use( new MagicHtmlMasterPageWrapper() )` using the middleware pattern.

We need to investigate how to add middleware. This should be done by considering how it is done in dotnet (aka dnx) so things does not brake in future versions.

We will keep a deprecated Page class that uses the new tech for backwards compatibility. Starcounter should **NOT** be allowed to reference this class anywhere in its code base.

**Comment by Starcounter-Jack** [comment]
@per-samuelsson @warpech @miyconst 
WDYT?

**Comment by Starcounter-Jack** [comment]
```c#

class MyApp 
{
    static void Main() {
         Handle.GET("/person", (Request r) => {
              if (r.Accept.Html) {
                   return Self.GET("/person.html");
              }
              return new PersonPage();
         });
    }
}

public partial class PersonPage : Json {
}
```

personpage.json
```
{
   FirstName:"",
   LastName:""
}
```

**Comment by miyconst** [comment]
Personally, I don't like the idea.

Such example:

A WebPage to edit user profile has two states: Editable, Readonly.
The Page is loaded to the client with puppet, mime type: `application/json`.

On the client side `<starcounter-include>` or any other component imports `Html`, which is a property of the view-model. User presses button Edit, and page switches its state from Readonly to Editable by replacing value of `Html` property.

This possibility would be gone, if we go with the approach proposed above. I know, there are lots of workarounds to get same behavior, but this one will gone.

One more thing, Starcounter applications are mostly Web based, and this code:

```cs
if (r.Accept.Html) {
    // ...
}
```

Will be copied almost for every handler. In the end every developer will have own copy of "Page" class. Which will result in one more file for our sample apps.

**Comment by per-samuelsson** [comment]
Rather than

```C#
Handle.GET("/person", (Request r) => {
  if (r.Accept.Html) { 
    return Self.GET("/person.html");
  }
  return new PersonPage();
});
```

I'd prefer something like

```C#
Handle.GET("/person", => {
  return new PersonPage().WithHtml("/person.html");
  // And overloaded like for lazy loading
  // return new PersonPage().WithHtml(file => "/person.html");
});
```

as our default pattern. I like the fact that some outside authority decide what format to use, I just provide the alternatives. If the request come for JSON, the authority use the plain JSON. If it comes for HTML, it invokes that provider.

As an *alternative*, we provide a pattern that allows custom MIME type switching, for more complex scenarios. Lets try not to include the `Request` object in our default patterns.

**Comment by per-samuelsson** [comment]
Background: #3165 and #3200.

**Comment by Starcounter-Jack** [comment]
>>  In the end every developer will have own copy of "Page" class. Which will result in one more file for our sample apps.

I think so to. But then there will not be as much magic. We will not have an extra class in our samples because we will do it with an if statement.

**Comment by Starcounter-Jack** [comment]
>> On the client side `<starcounter-include>` or any other component imports `Html`, which is a property of the view-model. User presses button Edit, and page switches its state from Readonly to Editable by replacing value of `Html` property.

This one is trickier. The user is free to implement a Html property but the question is how to trigger puppet to reload the HTML.

**Comment by Starcounter-Jack** [comment]
The question is wether we want hidden magic or not.

**Comment by miyconst** [comment]
> ... because we will do it with an if statement.

I would do that only if instructed so. Copy pasting any code is always a bad practice.

> But then there will not be as much magic.

The magic itself is not a bad thing as long as it works for us. Any framework requires some magic. You keep using Mac OS with all its magic and don't switch to Linux, which is the most magic less system.

**Comment by Starcounter-Jack** [comment]
>> Copy pasting any code is always a bad practice.

Developer already knows how to create base classes. They will do that. The importance of samples is NOT to promote teachings about these topics. They should learn generic practises in larger samples or in other literature. These are hello world samples to explain Starcounter.

We will ***NOT*** do the J2EE/Microsoft approach of having dozens of files in small samples. We will have minimal apps consisting of three files:

product.html
product.json
product.json.cs

**Comment by miyconst** [comment]
As a developer, I try to follow examples I find on official sources. If an example shows, that I have to use an `if` to check mime type - I will do so.. well.. a couple of time. Then I will find it stupid and will add another class, which would does that automatically.

The custom class will do a common thing for almost every Starcounter application. Each new developer working on the solution would have to learn, how me as the project author implemented this feature in my case. 

On other side, when this is a Starcounter feature, everyone gets to know it from the very beginning, don't have to reinvent the wheel and don't have to share "unique class for common feature" with others.

Another argument. For whatever reason we decide to change the API and the `if (r.Accept.Html)` is no longer supported. What then? We have to modify every sample app and tutorial.

Another argument. Our handlers can magically return `string`, `number`, and `json`. Why we don't remove this magic as well?

Another argument. SocietyObjects and Simplified libraries contain huge number of classes. We do not remove this libraries, but try to make user experience as smooth as possible by providing mappers. So the users don't have to deal with the libraries on their own.

**Comment by warpech** [comment]
While we are at it, it's worth mentioning that `Partial` class now not only provides implicit standalone mode, but also [implicit session](https://github.com/Starcounter/Starcounter/blob/develop/src/Starcounter/Partial.cs#L51) (which is needed to handle patching the partial).

Thanks to this, we have managed to shrink this:

```cs
            Handle.GET("/SomeApp/Master", () => {
                Session session = Session.Current;

                if (session != null && session.Data != null)
                    return session.Data;

                var master = new MasterPage();

                if (master == null) {
                    master = new Session(SessionOptions.PatchVersioning);
                    master.Html = "/SomeApp/MasterPage.html";
                }
                else {
                    master.Html = "/SomeApp/LauncherWrapperPage.html";
                }

                standalone.Session = session;
                return standalone;
            });
```

To:

```cs
            Handle.GET("/SomeApp/Master", () => {
                Session session = Session.Current;

                if (session != null && session.Data != null)
                    return session.Data;

                var master = new Master();
                master.Html = "/SomeApp/MasterPage.html";
                return master;
            });
```

This is simpler code with admittedly more magic. Didn't we want that with Barcodes2? 

I like `Partial` more than the lack of it, but I wholeheartedly agree that it could be solved better.

@per-samuelsson and @Starcounter-Jack have a point that `Partial` could be slightly more explicit than right now. Both examples from https://github.com/Starcounter/Starcounter/issues/3202#issuecomment-165733091 seem nice and understandable to me.

**Comment by warpech** [comment]
> User presses button Edit, and page switches its state from Readonly to Editable by replacing value of `Html` property.

This is technically correct, but realistically none of sample apps nor known customer apps use do this. We should not optimize for unexisting problems. 

These patterns are used almost exclusively to reload partials:

- replacing a subtree partial with another partial (using `Self.GET`)
- replacing parent partial with another partial (using `<puppet-redirect>`)
- replacing data of the current partial without reloading Html

**Comment by per-samuelsson** [comment]
Just want to make clear I really like this

> While we are at it, it's worth mentioning that `Partial` class now not only provides implicit standalone mode, but also [implicit session](https://github.com/Starcounter/Starcounter/blob/develop/src/Starcounter/Partial.cs#L51) (which is needed to handle patching the partial).

and agree with @warpech on this

> I like `Partial` more than the lack of it, but I wholeheartedly agree that it could be solved better.

There is always a balance between "magic" and explicitness. Some like magic, some don't. And there are grey zones.

And when on the subject: we shouldn't mix magic with support for convention patterns. For me, magic is more often bad when its not based on conventions. Take a system as **Maven** for example. One could claim there's a lot of magic in it. But to me, its not. Its just that they use the fact that people tend to like to follow good conventions. Thats a direction I'd like to see us move towards.

**Comment by Starcounter-Jack** [comment]
If we move html merging and standalone mode to middlewarr, we could have both. I.e. Partial and Page would be sugaring.

**Comment by per-samuelsson** [comment]
> It feels like overkill to have both Partial and Page though.

I agree, and I think that never was the plan: the plan was to *replace* `Page` with `Partial`. They just exist in parallel b/c `Partial` was really new, and we wanted to test it out before deprecating `Page`.

Correct me if I'm wrong, @warpech.

**Comment by per-samuelsson** [comment]
> If we move html merging and standalone mode to middleware, we could have both.

Yes, this is what I want too. `Partial` would be sugaring.

For me, I vote we implement a pattern where you can return something for a request that allow the *code host* to pick the most appropriate "response renderer" based on the accepted MIME type.

I think it's fine that we promote that what you normally return is either JSON or HTML. Basically, our message goes 

> "the convention is you return JSON or HMTL; if you want to return something else, we support that, but there is a few extra lines for you to write"

Why? Because these two data formats will be the absolute majority in all our samples, and the most common ones for our customers. Hence, we should make those two as easy as possible to return.

At the very core, when I write my handler for "/person", what I would want is a simple way to return JSON, *and a way* to specify how that JSON should be paired with a HTML / partial if the requested MIME type comes to that. And if I don't specify that last thing explicitly for my handler, registered middlewhere comes into play.

So, without middleware we support something like this:

```C#
Handle.GET("/myapp/person", () => {
  return new PersonJson().AndIfTheRequestIsTextHtmlPairThatJsonWith("/myapp/person.html");
}
```

By that, my handler support mime types JSON and HTML, and I don't have to show how to include middleware filtering. This is good for simple apps and samples.

In addition, we support like

```C#
Handle.GET("/myapp/person", () => {
  return new PersonJson();
}
```

and if the request is for `text/html`, the code host invokes a *user-supplied middleware* lookup, registered previously like

```C#
void Main() {

  Handle.RegisterHtmlResponseFor<Json>(Request request, Json json, string path) {
    Assert(request.Accept["text/html"]);
    Assert(request.DoesNotAccept["application/json"]);  // Because if it did, we wouldnt get here

    // Here, we provide a way to structure by convention, and enable that by this registring.
    // The convention could be like if path == "/myapp/person", the looking goes
    // "[resourcedirs]\myapp\person.html". This is a nice balance: we support a convention,
    // but the application must enable it by registering this middleware thingy.
    var html = Resources.GetHtmlByDefaultConvention(path);

    if (html != null) return new Partial(html, json);

    // At the root of the chain, we have one that return a HTML like 404 and describe what
    // resorce was requested, and that no HTML could be found for that - AND provide a link
    // to a tutorial showing how to enable HTML using any of the patterns in this post.
    return NextMiddleware(...);
  }

  Handle.GET("/myapp/person", () => {
    // This now end up in the above, if the request comes `text/html´
    return new PersonJson();
  }
}
```

**Comment by per-samuelsson** [comment]
We have a work issue for this now, and it's in the loop. I'm responsible for #3251, so I think we can close this one down once we nail that one. I'll therefor reassign this one to me as well.

**Comment by per-samuelsson** [comment]
Done.

**Comment by warpech** [comment]
The usage of the class was mostly removed in sample apps, but it is still present in docs and Starcounter (https://github.com/Starcounter/Starcounter/blob/develop/src/Starcounter/Page.cs), so why closing as Done?

**Comment by per-samuelsson** [comment]
> The usage of the class was mostly removed in sample apps, but it is still present in docs and Starcounter (https://github.com/Starcounter/Starcounter/blob/develop/src/Starcounter/Page.cs), so why closing as Done?

Well, so that we could have one less open issue I guess? 😉 

I thought we decided to keep `Page` and `Partial` classes for backwards compatibility, but bump middleware as the recommended approach. I probably understood something wrong. What would you say is still needed to consider this "done"?

**Comment by warpech** [comment]
You are right, let's keep backwards compatibility for a while. But then let's obsolete `Page` and `Partial` classes.

In docs, rewrite the pages to not use the `Page` class:

- https://docs.starcounter.io/v/Develop/guides/mapping-and-blending/uri-mapping/
- https://github.com/Starcounter/Docs/tree/Develop/guides/network/using-middleware
- https://docs.starcounter.io/v/Develop/guides/puppet-web-apps/more-on-partials/
- https://docs.starcounter.io/v/Develop/guides/typed-json/callback-methods-of-starcounter-js/

**Comment by Mackiovello** [comment]
> In docs, rewrite the pages to not use the Page class:

Fixed in [8300a](https://github.com/Starcounter/Docs/commit/8300a5c4424549b8dd4562d15b816902cb278da0) on all branches except 2.1.177 for the pages that are in the docs. The using-middleware page is not updated because it's obsolete.

**Comment by warpech** [comment]
```
return new Json() { Html = "/People/viewmodels/Menu.html" };
```

^ This is kind of a mental shortcut, because Json class does not have a Html property unless you explicitly declare it in a derived class.

Me and @tomalec started thinking about removing the term "partial" from our docs completely: https://github.com/Starcounter/TechComm/issues/286
